JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java

```diff
@@ -0,0 +1,318 @@
+		}
+	}
+
+	private CandidateProgram mutate(final CandidateProgram program) {
+		CandidateProgram mutated = program.clone();
+		if (this.rng.nextDouble() < this.getConfig().getMutationRate()) {
+			WhighamMutation mutation = new WhighamMutation(this.rng);
+			mutated = mutation.mutate(mutated);
+		}
+		return mutated;
+	}
+
+	private CandidateProgram tournament(final List<CandidateProgram> population) {
+		List<CandidateProgram> candidates = new ArrayList<CandidateProgram>(population);
+		Collections.shuffle(candidates, new Random(this.rng.nextLong()));
+		List<CandidateProgram> tournamentCandidates = IntStream.range(0, this.getConfig().getTournamentSize()).mapToObj(x -> candidates.get(x)).collect(Collectors.toList());
+		Collections.sort(tournamentCandidates);
+		return tournamentCandidates.get(0);
+	}
+
+	private void evaluate(final List<CandidateProgram> population) throws InterruptedException {
+		ExecutorService pool = Executors.newFixedThreadPool(this.getConfig().cpus());
+		Semaphore semaphore = new Semaphore(0);
+
+		AtomicBoolean interrupted = new AtomicBoolean(false);
+
+		for (CandidateProgram individual : population) {
+			if (this.cacheMap.containsKey(individual.toString())) {
+				((GRCandidateProgram) individual).setFitnessValue(this.cacheMap.get(individual.toString()));
+				semaphore.release();
+			} else {
+				Runnable evaluateTask = new Runnable() {
+					@Override
+					public void run() {
+						try {
+							if (Thread.interrupted() || interrupted.get()) {
+								semaphore.release();
+							}
+							ComponentInstance ci = GrammarBasedGeneticProgramming.this.converter.grammarStringToComponentInstance(individual.toString());
+							GRCandidateProgram realInd = ((GRCandidateProgram) individual);
+							try {
+								double fitnessValue = GrammarBasedGeneticProgramming.this.evaluator.evaluate(ci);
+								if (GrammarBasedGeneticProgramming.this.updateBestSeenSolution(new GGPSolutionCandidate(ci, fitnessValue))) {
+									GrammarBasedGeneticProgramming.this.earlyStoppingCounter.set(0);
+								}
+								realInd.setFitnessValue(fitnessValue);
+							} catch (ObjectEvaluationFailedException | InterruptedException e) {
+								realInd.setFitnessValue(10000.0);
+							} finally {
+								semaphore.release();
+							}
+						} catch (Throwable e) {
+							System.err.println(individual.toString());
+							e.printStackTrace();
+							semaphore.release();
+						}
+					}
+				};
+				pool.submit(evaluateTask);
+			}
+		}
+		try {
+			semaphore.acquire(population.size());
+		} catch (InterruptedException e) {
+			interrupted.set(true);
+			pool.shutdownNow();
+			throw e;
+		}
+		population.stream().forEach(x -> {
+			try {
+				this.cacheMap.put(x.toString(), x.getFitness());
+			} catch (Exception e) {
+				// could not cache fitness => nullpointer exception must have occurred, so ignore the candidate and do not put it into the cache.

```

Comment:
maybe log something here?

Reply:
No if the evaluation failed, there has already been a log output in the evaluation block. So at this point it is only about caching individual's fitness values for convenience (to not re-evaluate the same individuals over and over again).